### PR TITLE
feat(shared): add 12 net-new DCR MCP catalog entries fixes NV-7959

### PR DESCRIPTION
**Pull request by djabarovgeorge**
## New DCR MCP servers (12)

Stacked on #11435 (`fix/mcp-oauth-dcr-polish`) — catalog-only follow-up after OAuth discovery polish lands.

| Name | Catalog ID |
|------|------------|
| Axiom | `axiom` |
| Better Stack | `better-stack` |
| Buildkite | `buildkite` |
| GitLab | `gitlab` |
| Grafana Cloud | `grafana` |
| LaunchDarkly | `launchdarkly` |
| New Relic | `new-relic` |
| Pulumi | `pulumi` |
| Railway | `railway` |
| Replicate | `replicate` |
| Semgrep | `semgrep` |
| Sourcegraph | `sourcegraph` |

---

## Summary

- Add 12 net-new MCP catalog entries with `oauth.mode: dcr`, each vetted via live PRM/DCR probes (see `.cursor/skills/onboard-dcr-mcp/`).
- Depends on #11435 for issuer-relaxation and DCR flow polish (e.g. New Relic sibling-subdomain, GitLab).

Linear: [NV-7959](https://linear.app/novu/issue/NV-7959/expand-mcp-catalog-dcr-onboarding-batch-oauth-discovery-polish)

## Test plan

- [ ] `pnpm --filter @novu/shared build`
- [ ] Spot-check OAuth connect for GitLab + New Relic + one other in staging after #11435 is deployed
- [ ] Confirm all 12 appear in agent MCP picker with DCR wiring

Made with [Cursor](https://cursor.com)

## What changed

This PR adds 12 new MCP catalog entries with Dynamic Client Registration (DCR) authentication support. The entries are for Axiom, Better Stack, Buildkite, GitLab, Grafana Cloud, LaunchDarkly, New Relic, Pulumi, Railway, Replicate, Semgrep, and Sourcegraph. Each entry includes the provider's name, description, URL, and DCR oauth configuration. The changes build on PR `#11435`, which introduced issuer-relaxation and DCR flow improvements required for these providers.

## Affected areas

**shared**: The `mcp-servers.ts` catalog was extended with 12 new `McpServer` entries configured for DCR authentication, adding them to both popular and "all others" sections.

**api**: Two test suites were enhanced. The `mcp-oauth-discovery.service.spec.ts` now includes tests validating issuer pattern scenarios (delegated-issuer, parent-domain, sibling-subdomain, MCP gateway patterns), plus a test confirming issuer normalization with trailing slashes. The `mcp-oauth-issuer-match.spec.ts` adds negative test cases verifying that issuer-relaxation patterns are rejected when the OAuth token endpoint moves to an unauthorized domain.

## Testing

New unit tests were added to validate issuer-discovery edge cases: delegated issuer on product origin, parent-domain issuers, sibling-subdomain acceptance, MCP gateway patterns, and trailing-slash normalization. Tests also verify that invalid endpoint-to-domain binding is rejected. Manual verification is planned for staging to confirm OAuth connect works for GitLab, New Relic, and one additional provider after PR `#11435` is deployed, and to verify all 12 entries appear in the agent MCP picker with correct DCR wiring.

<h3>Greptile Summary</h3>

This PR adds 12 net-new MCP catalog entries with `oauth.mode: dcr` (Dynamic Client Registration) to `mcp-servers.ts`, covering providers like GitLab, Grafana Cloud, New Relic, LaunchDarkly, Axiom, and others. Accompanying tests extend the OAuth discovery and issuer-match specs with acceptance and rejection scenarios for delegated-issuer, parent-domain, sibling-subdomain, and MCP gateway patterns.

- **Catalog additions** (`mcp-servers.ts`): 12 new `McpServer` entries spread across the popular and all-others sections, each with a vetted DCR OAuth config.
- **Discovery service tests** (`mcp-oauth-discovery.service.spec.ts`): New positive tests validate the service processes Clerk delegated-issuer, Vercel parent-domain, PlanetScale MCP gateway, New Relic sibling-subdomain, and Monte Carlo trailing-slash patterns; one new rejection test verifies the service throws `mcp_no_as_metadata` when an OAuth endpoint moves to an unauthorized domain.
- **Issuer-match tests** (`mcp-oauth-issuer-match.spec.ts`): Three new negative cases confirm `isAcceptableIssuerMatch` rejects each pattern variant when the token endpoint is swapped to `evil.example`.

<h3>Confidence Score: 5/5</h3>

Safe to merge; purely additive catalog entries and tests with no changes to runtime logic.

All 12 catalog additions follow the established McpServer schema, and the PR description notes each endpoint was vetted via live PRM/DCR probes. The new tests correctly exercise both acceptance and rejection paths for each issuer-relaxation pattern, and the negative cases in both spec files explicitly guard against adversarial token-endpoint substitution.

No files require special attention.

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| packages/shared/src/consts/providers/mcp-servers.ts | 12 new DCR MCP catalog entries added with correct structure; all entries follow the existing McpServer schema pattern. |
| apps/api/src/app/agents/mcp/oauth/mcp-oauth-discovery.service.spec.ts | Six new test cases added covering acceptance and rejection of issuer-relaxation patterns; consistent with the existing try/catch test style used throughout the file. |
| apps/api/src/app/agents/mcp/oauth/mcp-oauth-issuer-match.spec.ts | Three new negative test cases added verifying that delegated-issuer, parent-domain, and sibling-subdomain relaxations are rejected when the token endpoint moves to an untrusted domain. |

</details>

</details>

<h3>Flowchart</h3>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[User selects DCR MCP provider] --> B[Load catalog entry from mcp-servers.ts]
    B --> C[discoverAuthorizationServer\ncatalog URL]
    C --> D{PRM probe\n/.well-known/oauth-protected-resource}
    D -->|200 with authorization_servers| E[Fetch AS metadata]
    D -->|404 / no PRM| F[Try direct AS discovery\n/.well-known/oauth-authorization-server]
    E --> G[isAcceptableIssuerMatch\ndiscovery URL x issuer x endpoints]
    F --> G
    G -->|exact match| H[Proceed to DCR registration]
    G -->|relaxed match\nparent-domain / sibling / delegated| I{All OAuth endpoints\non trusted domain?}
    G -->|no match| J[throw mcp_no_as_metadata]
    I -->|yes| H
    I -->|evil.example in token endpoint| J
    H --> K[POST registration_endpoint\nDynamic Client Registration]
    K --> L[Store client_id + client_secret]
    L --> M[OAuth PKCE Authorization Code Flow]
```

<sub>Reviews (3): Last reviewed commit: ["test(api): add off-domain rejection case..."](https://github.com/novuhq/novu/commit/47e15d8509c46352c9ef8b2ca2295ffb9aa40ffd) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=35383017)</sub>

### CODE REVIEW

File: packages/shared/src/consts/providers/mcp-servers.ts [hunk 1]
```diff
@@ -302,6 +311,24 @@ export const MCP_SERVERS: McpServer[] = [
...
+    url: 'https://mcp.grafana.com/mcp',
+    category: 'code',
+    popular: true,
+    oauth: { mode: McpConnectionAuthModeEnum.Dcr },
+  },
+  {
+    id: 'new-relic',
+    name: 'New Relic',
+    description: 'Query metrics, logs, traces, and alerts from New Relic using natural language.',
+    url: 'https://mcp.newrelic.com/mcp/',
```
[the remark this answers is not in the document]
Comment (djabarovgeorge) [reply packages/shared/src/consts/providers/mcp-servers.ts:326]: Fixed — removed the trailing slash so the URL is `https://mcp.newrelic.com/mcp`, consistent with other catalog entries. `buildPrmCandidates` already strips trailing slashes before building PRM paths, but the catalog entry should not carry the slash.

File: packages/shared/src/consts/providers/mcp-servers.ts [hunk 2]
```diff
@@ -574,6 +619,15 @@ export const MCP_SERVERS: McpServer[] = [
     popular: false,
     oauth: { mode: McpConnectionAuthModeEnum.ProviderManaged },
   },
+  {
+    id: 'better-stack',
+    name: 'Better Stack',
+    description: 'Query uptime monitors, incidents, logs, metrics, and dashboards from Better Stack.',
+    url: 'https://mcp.betterstack.com',
```
[the remark this answers is not in the document]
Comment (djabarovgeorge) [reply packages/shared/src/consts/providers/mcp-servers.ts:625]: Keeping `https://mcp.betterstack.com` (no `/mcp` path). Live PRM probe: `GET https://mcp.betterstack.com/.well-known/oauth-protected-resource` → 200 with `authorization_servers: ["https://betterstack.com"]`; `…/mcp/.well-known/oauth-protected-resource` → 404. The MCP resource is the host root per their PRM `resource` field.

File: apps/api/src/app/agents/mcp/oauth/dcr-provider-strategies/__fixtures__/_shared/replay.ts [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+function readJsonFile<T>(filePath: string): T {
+  return JSON.parse(readFileSync(filePath, 'utf8')) as T;
+}
+
+function readOptionalJsonFile<T>(filePath: string): T | undefined {
+  try {
+    return readJsonFile<T>(filePath);
+  } catch {
+    return undefined;
+  }
```
[the remark this answers is not in the document]
Comment (djabarovgeorge) [reply apps/api/src/app/agents/mcp/oauth/dcr-provider-strategies/__fixtures__/_shared/replay.ts]: Resolved: \`dcr-provider-strategies/\` was removed in d06c518 (superseded by \`token-exchange-outcome.ts\` on #11435).

File: apps/api/src/app/agents/mcp/oauth/dcr-provider-strategies/resolve-default-dcr-token-exchange-outcome.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
+import { mapTokenExchangeErrorCode } from '../mcp-oauth-callback/mcp-oauth-callback.usecase';
```
[the remark this answers is not in the document]
Comment (djabarovgeorge) [reply apps/api/src/app/agents/mcp/oauth/dcr-provider-strategies/resolve-default-dcr-token-exchange-outcome.ts]: Resolved: \`resolve-default-dcr-token-exchange-outcome.ts\` was removed with the strategies module; logic lives in \`token-exchange-outcome.ts\`.

File: apps/api/src/app/agents/mcp/oauth/mcp-oauth-discovery.service.spec.ts [hunk 5]
```diff
@@ -267,6 +267,122 @@ describe('McpOAuthDiscoveryService', () => {
...
+      );
+
+      const md = await service.discoverAuthorizationServer('https://mcp.newrelic.com');
+
+      expect(md.issuer).to.equal('https://login.newrelic.com');
+      expect(md.authorizationEndpoint).to.equal('https://login.newrelic.com/login');
+      expect(md.tokenEndpoint).to.equal('https://mcp.newrelic.com/oauth2/token');
+      expect(md.registrationEndpoint).to.equal('https://mcp.newrelic.com/register');
+    });
+
```
[the remark this answers is not in the document]
Comment (djabarovgeorge) [reply apps/api/src/app/agents/mcp/oauth/mcp-oauth-discovery.service.spec.ts:347]: Addressed in \`mcp-oauth-issuer-match.spec.ts\` — added off-domain rejection tests for delegated-issuer, parent-domain, and sibling-subdomain patterns (issuer logic moved out of discovery.service.spec).
